concent_api/core/views.py

```diff
@@ -0,0 +1,223 @@
+                decoded_message_to_compute['timestamp'] = current_time
+                return decoded_message_to_compute
+            return decoded_message_data
+        return HttpResponse("", status = 204)
+
+    if decoded_message_data['message_task_to_compute']['deadline'] + settings.CONCENT_MESSAGING_TIME <= current_time <= decoded_message_data['message_task_to_compute']['deadline'] + 2 * settings.CONCENT_MESSAGING_TIME:
+        decoded_message_data['type']      = "MessageAckReportComputedTask"
+        decoded_message_data['timestamp'] = current_time
+        return decoded_message_data
+
+    if current_time <= decoded_message_data['message_task_to_compute']['deadline'] + settings.CONCENT_MESSAGING_TIME:
+        return decoded_message_data
+
+    return HttpResponse("", status = 204)
+
+
+@api_view
+@require_POST
+def receive_out_of_band(_request, _message):
+    last_task_message = Message.objects.order_by('id').last()
+    if last_task_message is None:
+        return None
+
+    raw_last_task_message     = last_task_message.data.tobytes()
+    decoded_last_task_message = json.loads(raw_last_task_message.decode('utf-8'))
+    current_time              = int(datetime.datetime.now().timestamp())
+    message_verdict           = {
+        "type":                               "MessageVerdictReportComputedTask",
+        "timestamp":                          current_time,
+        "message_force_report_computed_task": {},
+        "message_ack_report_computed_task":   {
+            "type":      "MessageAckReportComputedTask",
+            "timestamp": current_time
+        }
+    }
+
+    if decoded_last_task_message['type'] == "MessageForceReportComputedTask":
+        task_deadline = decoded_last_task_message['message_task_to_compute']['deadline']
+        if task_deadline + settings.CONCENT_MESSAGING_TIME <= current_time:
+            message_verdict['message_force_report_computed_task'] = decoded_last_task_message
+            return message_verdict
+
+    if decoded_last_task_message['type'] == "MessageRejectReportComputedTask":
+        if decoded_last_task_message['message_cannot_commpute_task']['reason'] == "deadline-exceeded":
+            rejected_task_id                                      = decoded_last_task_message['message_cannot_commpute_task']['task_id']
+            message_to_compute                                    = Message.objects.get(type = 'MessageForceReportComputedTask', task_id = rejected_task_id)
+            raw_message_to_compute                                = message_to_compute.data.tobytes()
+            decoded_message_to_compute                            = json.loads(raw_message_to_compute.decode('utf-8'))
+            message_verdict['message_force_report_computed_task'] = decoded_message_to_compute
+            return message_verdict
+
+    return HttpResponse("", status = 204)
+
+
+def validate_message(data):
+    if 'type' not in data:
```

Comment:
To i kolejne funkcje moim zdaniem powinny nie być w pliku z widokami tylko w modelach, model powinien móc walidować sam siebie pod kątem poprawności danych.

Reply:
Yeah, I completely agree.

Not necessarily in the models though. `golem-messages` has some (limited) validation for its messages. We'll need it to be more verbose about what it finds though. Currently it just raises `AttributeErrors` which is hard to reliably discern from a genuine bug in the code.

But that's another thing that I'm ignoring for the time being because it would disrupt @kbeker's work too much. We'll get back to this when he's done with his current branch.

Reply:
:+1:

Reply:
Maybe we can keep track of those small upgrades that we are planning to do in pivotal ?

Reply:
I think github issues is a better place. Pivotal is more for high-level stuff and only for concent. golem-messages is maintained independently by the core Golem team.

I'm actually going to report it in `golem-messages` on Github but I'm currently overloaded with work so it's still sitting on my TODO list and waiting for its turn.